### PR TITLE
[TTAHUB-519] Grant Program Personnel table and import

### PR DESCRIPTION
**Pull request by AdamAdHocTeam**
## Description of change

This modifies the existing 'updateGrantsRecipients' job to also bring in program personnel associated with that grant. If a personnel for a specific grant and role has changed (first and last name). We mark the old as no longer active and set the new personnel as active.

Per the jira ticket we bring in the following roles:

- auth_official_contact
- ceo
- policy_council
- director for HS or director EHS

with the following fields: 

- prefix (Mr./Ms., etc...)
- first_name
- last_name
- suffix (Only example I saw was "CPA")
- title
- email

## How to test

Review the code and run the test to make sure all data is created correctly.

## Issue(s)

* https://ocio-jira.acf.hhs.gov/browse/TTAHUB-519

### Every PR

- [x] Meets issue criteria
- [x] JIRA ticket status updated
- [x] Code is meaningfully tested
- [x] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [ ] API Documentation updated
- [ ] Boundary diagram updated
- [ ] Logical Data Model updated
- [ ] [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions
- [ ] UI review complete

### Production Deploy

- [ ] Staging smoke test completed

### After merge/deploy

- [x] Update JIRA ticket status

### CODE REVIEW

File: src/models/program.js [hunk 1]
```diff
@@ -1,34 +1,35 @@
...
+      autoIncrement: false,
+    },
+    grantId: {
+      type: DataTypes.INTEGER,
+      allowNull: false,
+    },
+    programType: DataTypes.STRING,
+    startYear: DataTypes.STRING,
+    startDate: DataTypes.STRING,
+    endDate: DataTypes.STRING,
```
Comment (GarrettEHill) [inline src/models/program.js:27]: Can these be some form of date column

File: docs/logical_data_model.puml [hunk 2]
```diff
@@ -466,6 +466,24 @@ class Permissions{
  * scopeId : integer : REFERENCES "Scopes".id
 }
 
+class ProgramPersonnel{
+ * id : integer : <generated>
+ * grantId : integer
+ * programId : integer
+ * role : varchar(255)
```
[the remark this answers is not in the document]
Comment (AdamAdHocTeam) [reply docs/logical_data_model.puml:473]: This role is specific to programs. 

examples would be: auth_official_contact, director, and ceo.

File: src/models/program.js (hunk 1 above)
Comment (AdamAdHocTeam) [reply src/models/program.js:27]: I think it comes in as a string.

Comment (AdamAdHocTeam) [reply src/models/program.js:27]: I'm just realizing these are not part of this PR.

File: docs/logical_data_model.puml [hunk 3]
```diff
@@ -466,6 +466,24 @@ class Permissions{
...
+ * active : boolean
+ prefix : varchar(255)
+ firstName : varchar(255)
+ lastName : varchar(255)
+ suffix : varchar(255)
+ title : varchar(255)
+ email : varchar(255)
```
[the remark this answers is not in the document]
Comment (AdamAdHocTeam) [reply docs/logical_data_model.puml:480]: I spoke to Angela the earliest these could become real users would be 5 yrs (or so) down the road. We can add a user id link column down the road if we need it.

File: src/lib/updateGrantsRecipients.js [hunk 4]
```diff
@@ -168,11 +235,22 @@ export async function processFiles(hashSumHex) {
...
+        // Get grant personnel.
+        programPersonnel: await getProgramPersonnel(
+          parseInt(grantAgencyMap[program.grant_agency_id], 10),
+          parseInt(program.grant_program_id, 10),
+          program,
+        ),
+      })));
+
+      // Extract an array of all grant personnel to update.
+      const programPersonnel = programsForDb.map((p) => p.programPersonnel).flat();
```
Comment (nvms) [inline src/lib/updateGrantsRecipients.js:247]: not suggesting a change, just an FYI that there's a `flatMap` operator that does this :upside_down_face:

e.g. `programsForDb.flatMap(({ programPersonnel }) => programPersonnel);`

File: src/constants.js [hunk 5]
```diff
@@ -33,6 +33,13 @@ export const GOAL_STATUS = {
   CLOSED: 'Closed',
 };
 
+export const GRANT_PERSONNEL_ROLES = [
```
Comment (nvms) [inline src/constants.js:36]: do these need to be added to `@ttahub/common` as well, or will they only ever be used by this update job?

**Review by nvms — COMMENTED** [review]
minor comments

small bugfix (the out-of-bounds possibility) that should be fixed

otherwise looks good!

File: src/models/program.js [hunk 6]
```diff
@@ -1,34 +1,35 @@
...
-    endDate: DataTypes.STRING,
-    status: DataTypes.STRING,
-    name: DataTypes.STRING,
-  }, {
-    sequelize,
-    modelName: 'Program',
-  });
-  return Program;
-};
+const {
```
Comment (thewatermethod) [inline src/models/program.js:1]: weird, was this not saved with the LF endings? It makes it challenging to resolve merge conflicts (should there be any) when we have these different file endings floating around

File: src/lib/updateGrantsRecipients.js [hunk 7]
```diff
@@ -223,6 +302,15 @@ export async function processFiles(hashSumHex) {
           transaction,
         },
       );
+
+      // Update grant personnel.
+      await ProgramPersonnel.unscoped().bulkCreate(
```
Comment (thewatermethod) [inline src/lib/updateGrantsRecipients.js:307]: Is there a reason to use unscoped here? If we ever decide to add a default scope or something this may cause some annoyance

**Comment by thewatermethod** [comment]
Looks pretty straightforward, just have to fix the merge conflicts and apply Jon's changes

File: src/lib/updateGrantsRecipients.js [hunk 8]
```diff
@@ -27,6 +31,69 @@ function combineNames(firstName, lastName) {
   return joinedName === '' ? null : joinedName;
 }
 
+function getPersonnelFields(role, field, program) {
+  return typeof program[`${role}_${field}`] === 'object' ? null : program[`${role}_${field}`];
+}
+
+async function getProgramPersonnel(grantId, programId, program) {
+  const programPersonnelArray = [];
+  for (let i = 0; i <= GRANT_PERSONNEL_ROLES.length; i += 1) {
```
[the remark this answers is not in the document]
Comment (AdamAdHocTeam) [reply src/lib/updateGrantsRecipients.js:40]: oops good catch.

[the remark this answers is not in the document]
Comment (AdamAdHocTeam) [reply src/lib/updateGrantsRecipients.js:40]: good catch will make it <

File: src/lib/updateGrantsRecipients.js (hunk 4 above)
Comment (AdamAdHocTeam) [reply src/lib/updateGrantsRecipients.js:247]: Thought I tried that with no luck. Let me try again.

File: src/constants.js (hunk 5 above)
Comment (AdamAdHocTeam) [reply src/constants.js:36]: I think its fine to leave these here for now as they are only used by the BE job.

File: src/lib/updateGrantsRecipients.test.js [hunk 9]
```diff
@@ -14,6 +14,8 @@ jest.mock('./fileUtils', () => ({
   fileHash: () => 'hash',
 }));
 
+jest.mock('bull');
```
[the remark this answers is not in the document]
Comment (AdamAdHocTeam) [reply src/lib/updateGrantsRecipients.test.js:17]: I don't know why but mocking bull allows the test to run again in the same window without issue. I will remove for now.

File: src/lib/updateGrantsRecipients.js (hunk 7 above)
Comment (AdamAdHocTeam) [reply src/lib/updateGrantsRecipients.js:307]: I was following the same pattern as the other bulk update.

File: src/models/program.js (hunk 6 above)
Comment (AdamAdHocTeam) [reply src/models/program.js:1]: Not sure what I can do about this.

File: src/lib/updateGrantsRecipients.js [hunk 10]
```diff
@@ -27,6 +31,69 @@ function combineNames(firstName, lastName) {
   return joinedName === '' ? null : joinedName;
 }
 
+function getPersonnelFields(role, field, program) {
```
[the remark this answers is not in the document]
Comment (AdamAdHocTeam) [reply src/lib/updateGrantsRecipients.js]: This has been changed.

File: src/lib/updateGrantsRecipients.js (hunk 4 above)
Comment (AdamAdHocTeam) [reply src/lib/updateGrantsRecipients.js]: Yea it worked odd wonder what I did wrong the first time.

File: src/lib/updateGrantsRecipients.js (hunk 7 above)
Comment (thewatermethod) [reply src/lib/updateGrantsRecipients.js:311]: The other bulk update is for Grants, which have a default scope.

Comment (AdamAdHocTeam) [reply src/lib/updateGrantsRecipients.js]: I removed the .unscoped()
